---
title: "preprocessing_data_with_Cell_Ranger_"
output: html_document
---
## Cell Ranger
Cell Ranger is a set of analysis pipelines that process single-cell 
data to align reads and generate feature-barcode matrices.  To run Cell Ranger,
first create simple csv file with samplesheet with Lane,Sample,Index columns. 

## Run mkfastq to create fastq file
Run the following command
$ cellranger mkfastq --id= \
                   --run= \
                   --csv=
Here,
- id is the name of the folder to be created by mkfastq
- run is the pathway to Illumina BCL run folder
- csv is the pathway to a simple csv with lane, sample, and index columns.

## Run count to create count matrices
It generates single cell feature counts which can be used by Seurat for down
stream analysis. To run the count use the following command:
$ cellranger count --id= \
                   --transcriptome= \
                   --fastqs= \
                   --sample= \
                   --expect-cells= \
                   --localcores= \
                   --localmem=
Here,
transcriptome is the path to transcriptome reference
- fastqs is the path to fastq file generated by above step
- sample is the sample name as specified in the sample sheet supplied to mkfastq.
- expect-cells is the expected number of recovered cells. 
- localcores is the specified number of cores
- amount of memory(GB) specified to execute the pipeline