---
title: "Package ‘SeuratObject’"
output: learnr::tutorial
runtime: shiny_prerendered
---

Setup the Seurat Object

```{r setup, include=FALSE}
#install.packages("learnr")
#install.packages("dplyr")
#install.packages('Seurat')
#install.packages('patchwork')
#install.packages("Rcpp")

library(learnr)
library(dplyr)
library(Seurat)
library(patchwork)

knitr::opts_chunk$set(echo = FALSE)
```

The raw data includes information on 2,700 Peripheral Blood Mononuclear Cells (PBMC) freely available from 10X Genomics that were sequenced on the Illumina NextSeq 500 using Homo sapiens (human) genome assembly GRCh37 (hg19) as reference.

There are 13714 features across the 2700 samples within 1 single-cell RNA seq assay. 

```{r}

# Load the PBMC dataset
pbmc.data <- Read10X(data.dir = "~/Desktop/GDSSP/Codeathon_08.10/Creating-Rmarkdown-courses-to-teach-bioinformatic-analysis-pipelines/Descriptions/filtered_gene_bc_matrices/hg19/")

# Initialize the Seurat object with the raw (non-normalized data).
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc
```

Lets examine a few genes in the first thirty cells.

```{r}
pbmc.data[c("CD3D", "TCL1A", "MS4A1"), 1:30]
```

The . values in the matrix represent 0s (no molecules detected). Since most values in an scRNA-seq matrix are 0, Seurat uses a sparse-matrix representation whenever possible. This results in significant memory and speed savings for Drop-seq/inDrop/10x data.

The dense size of the array represents elements that have non-zero values
```{r}
dense.size <- object.size(as.matrix(pbmc.data))
dense.size
```

The sparse size of the array represents elements with a value of zero. 
```{r}
sparse.size <- object.size(pbmc.data)
sparse.size
```

The below ratio represents the amount of non-zero values in relation to zero values
```{r}
dense.size/sparse.size
```


Initialize the Seurat object with the raw (non-normalized data).  Keep all
genes expressed in >= 3 cells (~0.1% of the data). Keep all cells with at
least 200 detected genes

```{r}
pbmc <- CreateSeuratObject(raw.data = pbmc.data, min.cells = 3, min.genes = 200, 
    project = "10X_PBMC")
```


Standard pre-processing workflow
QC and selecting cells for further analysis

```{r}

#The PercentageFeatureSet function calculates the percentage of all counts that belong to a given set of features

#Function Usage
#PercentageFeatureSet(object,pattern = NULL,features = NULL,col.name = NULL,assay = NULL)

# The pattern attribute uses a regex pattern to match features against
# The "^" operator matches the beginning of a line or string. 
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats

pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")


# Show QC metrics for the first 5 cells
#meta.data refers to the information on the data, provides more context on the data
#orig.ident - refers to the data of ~3,000 PBMCs samples from a healthy donor
#nCount_RNA - total number of RNA molecules detected within a cell
#nFeature_RNA - total number of genes in each cell
#percent.mt - percent of mitochondrial genes within each cell
head(pbmc@meta.data, 5)

```

In the example below, we visualize QC metrics, and use these to filter cells.

We filter cells that have unique feature counts over 2,500 or less than 200
We filter cells that have >5% mitochondrial counts

```{r}

# Visualize QC metrics as a violin plot
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

```

